src/websockethandler.cpp

```diff
@@ -15,6 +15,60 @@ namespace {
 Logger logger(LOG_MAIN, "WebSocketHandler");
 }  // namespace
 
+ExponentialBackoffStrategy::ExponentialBackoffStrategy() {
+  MVPN_COUNT_CTOR(ExponentialBackoffStrategy);
+
+  m_retryTimer.setSingleShot(true);
+}
+
+#ifdef UNIT_TEST
+void ExponentialBackoffStrategy::testOverrideBaseRetryInterval(
+    int newInterval) {
+  m_baseInterval = newInterval;
+}
+
+void ExponentialBackoffStrategy::testOverrideMaxRetryInterval(int newInterval) {
+  m_maxInterval = newInterval;
+}
+#endif
+
+/**
+ * @brief Schedules the next attempt to execute a given function.
+ *
+ * Everytime a new attempt is scheduled the interval will be exponentially
+ * larger.
+ */
+int ExponentialBackoffStrategy::scheduleNextAttempt(
+    std::function<void()> action) {
+  int retryInterval = qPow(m_baseInterval, m_retryCounter);
+  if (retryInterval < m_maxInterval) {
+    m_retryCounter++;
+  }
+
+  QMetaObject::Connection* const connection = new QMetaObject::Connection;
+  *connection =
+      connect(&m_retryTimer, &QTimer::timeout, this, [action, connection]() {
```

Comment:
I wonder if this object should simply emit a signal and we can connect the signal to `open`. Would it be the same?
If we follow this QT pattern, we can connect the signal `GoodNameHere` of `m_backoffStrategy` in the CTOR of the `WebSocketHandler`. Then we can call: `schedule()` that will emit the signal, eventually.